File: SnipeMobile/Views/Maintenance/MaintenanceFormSheet.swift
import SwiftUI

struct MaintenanceFormSheet: View {
    @ObservedObject var apiClient: SnipeITAPIClient
    let assetId: Int
    let record: AssetMaintenance?
    var onSave: (Int) -> Void = { _ in }

    @Environment(\.dismiss) private var dismiss

    @State private var title: String = ""
    @State private var selectedType: String = "Maintenance"
    @State private var selectedTypeId: Int = 0
    @State private var selectedSupplierId: Int = 0
    @State private var cost: String = ""
    @State private var url: String = ""
    @State private var notes: String = ""
    @State private var isWarranty: Bool = false
    @State private var startDate: Date = Date()
    @State private var hasCompletionDate: Bool = false
    @State private var completionDate: Date = Date()
    @State private var selectedResponsibleId: Int = 0
    @State private var selectedImage: UIImage? = nil
    @State private var removeExistingImage: Bool = false
    @State private var isSaving: Bool = false
    @State private var showErrorAlert: Bool = false
    @State private var errorMessage: String = ""

    private let dateFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        f.timeZone = TimeZone(secondsFromGMT: 0)
        return f
    }()

    private var isEditing: Bool { record != nil }

    // empty list = older server without the maintenance-types endpoint
    private var usesTypeIds: Bool { !apiClient.maintenanceTypes.isEmpty }

    private var legacyTypeOptions: [String] {
        MaintenanceFormPickerSupport.legacyTypeOptions(
            selectedType: selectedType,
            recordType: record?.displayType
        )
    }

    private var typeIdPickerReady: Bool {
        MaintenanceFormPickerSupport.hasValidPickerTag(
            id: selectedTypeId,
            in: apiClient.maintenanceTypes.map(\.id)
        )
    }

    private var responsiblePickerReady: Bool {
        MaintenanceFormPickerSupport.hasValidPickerTag(
            id: selectedResponsibleId,
            in: apiClient.users.map(\.id)
        )
    }

    private var typeIsValid: Bool {
        usesTypeIds ? selectedTypeId != 0 : !selectedType.trimmingCharacters(in: .whitespaces).isEmpty
    }

    private var canSave: Bool {
        !title.trimmingCharacters(in: .whitespaces).isEmpty && typeIsValid && !isSaving
    }

    private var existingImageURL: URL? {
        guard let raw = record?.image?.trimmingCharacters(in: .whitespacesAndNewlines), !raw.isEmpty else {
            return nil
        }
        if let absolute = URL(string: raw), absolute.scheme != nil {
            return absolute
        }
        if raw.hasPrefix("/") {
            return URL(string: "\(apiClient.baseURL)\(raw)")
        }
        return nil
    }

    var body: some View {
        NavigationStack {
            Form {
                Section(header: Text(L10n.string("general"))) {
                    VStack(alignment: .leading, spacing: 4) {
                        Text(L10n.string("name"))
                            .font(.caption)
                            .foregroundColor(.secondary)
                        TextField(L10n.string("name"), text: $title)
                    }
                    if usesTypeIds {
                        if typeIdPickerReady {
                            Picker(L10n.string("maintenance_type"), selection: $selectedTypeId) {
                                ForEach(apiClient.maintenanceTypes) { type in
                                    Text(type.decodedName).tag(type.id)
                                }
                            }
                        } else {
                            Text(L10n.string("loading"))
                                .foregroundStyle(.secondary)
                        }
                    } else if !legacyTypeOptions.isEmpty {
                        Picker(L10n.string("maintenance_type"), selection: $selectedType) {
                            ForEach(legacyTypeOptions, id: \.self) { type in
                                Text(type).tag(type)
                            }
                        }
                    }
                }
                Section(header: Text(L10n.string("dates"))) {
                    DatePicker(L10n.string("start_date"), selection: $startDate, displayedComponents: .date)
                    Toggle(L10n.string("set_completion_date"), isOn: $hasCompletionDate)
                    if hasCompletionDate {
                        DatePicker(L10n.string("completion_date"), selection: $completionDate, displayedComponents: .date)
                    }
                }
                Section(header: Text(L10n.string("responsible_party"))) {
                    if responsiblePickerReady {
                        Picker(L10n.string("responsible_party"), selection: $selectedResponsibleId) {
                            ForEach(apiClient.users, id: \.id) { user in
                                Text(user.decodedName).tag(user.id)
                            }
                        }
                    } else {
                        Text(L10n.string("loading"))
                            .foregroundStyle(.secondary)
                    }
                }
                Section(header: Text(L10n.string("financial"))) {
                    VStack(alignment: .leading, spacing: 4) {
                        Text(L10n.string("cost"))
                            .font(.caption)
                            .foregroundColor(.secondary)
                        TextField(L10n.string("cost"), text: $cost)
                            .keyboardType(.decimalPad)
                    }
                    if !apiClient.suppliers.isEmpty {
                        Picker(L10n.string("supplier_optional"), selection: $selectedSupplierId) {
                            Text(L10n.string("none")).tag(0)
                            ForEach(apiClient.suppliers, id: \.id) { supplier in
                                Text(HTMLDecoder.decode(supplier.name)).tag(supplier.id)
                            }
                        }
                    }
                }
                Section {
                    Toggle(L10n.string("is_warranty"), isOn: $isWarranty)
                }
                Section(header: Text(L10n.string("url"))) {
                    TextField(L10n.string("url"), text: $url)
                        .keyboardType(.URL)
                        .textInputAutocapitalization(.never)
                        .disableAutocorrection(true)
                }
                AssetPhotoSection(
                    selectedImage: $selectedImage,
                    existingImageURL: existingImageURL,
                    removeExistingImage: $removeExistingImage
                )
                Section(header: Text(L10n.string("notes"))) {
                    TextEditor(text: $notes)
                        .frame(minHeight: 80)
                }
            }
            .navigationTitle(isEditing ? L10n.string("edit_maintenance") : L10n.string("add_maintenance"))
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button(L10n.string("cancel")) { dismiss() }
                }
                ToolbarItem(placement: .confirmationAction) {
                    Button(L10n.string("save")) {
                        Task { await save() }
                    }
                    .disabled(!canSave)
                }
            }
        }
        .onAppear {
            prefill()
            syncPickerSelections()
            if apiClient.suppliers.isEmpty {
                Task { await apiClient.fetchSuppliers() }
            }
            if apiClient.users.isEmpty {
                Task {
                    await apiClient.fetchUsers()
                    syncPickerSelections()
                }
            }
            if apiClient.maintenanceTypes.isEmpty {
                Task {
                    await apiClient.fetchMaintenanceTypes()
                    syncPickerSelections()
                }
            }
        }
        .onChange(of: apiClient.users.count) { _, _ in syncPickerSelections() }
        .onChange(of: apiClient.maintenanceTypes.count) { _, _ in syncPickerSelections() }
        .alert(L10n.string("error"), isPresented: $showErrorAlert) {
            Button(L10n.string("ok"), role: .cancel) {}
        } message: {
            Text(errorMessage)
        }
    }

    private func syncPickerSelections() {
        if usesTypeIds {
            MaintenanceFormPickerSupport.applyTypeIdSelection(
                selectedTypeId: &selectedTypeId,
                types: apiClient.maintenanceTypes,
                record: record
            )
        } else {
            MaintenanceFormPickerSupport.normalizeLegacyTypeSelection(
                selectedType: &selectedType,
                options: legacyTypeOptions
            )
        }
        MaintenanceFormPickerSupport.reconcileResponsibleSelection(
            selectedId: &selectedResponsibleId,
            users: apiClient.users,
            preferredId: record?.responsibleParty?.id,
            defaultUser: apiClient.defaultCheckoutUser
        )
    }

    private func prefill() {
        if let r = record {
            title = r.decodedTitle
            selectedType = r.assetMaintenanceType ?? r.maintenanceType ?? "Maintenance"
            selectedSupplierId = r.supplier?.id ?? 0
            cost = r.cost ?? ""
            url = r.url ?? ""
            notes = r.decodedNotes
            isWarranty = r.isWarranty
            if let partyId = r.responsibleParty?.id {
                selectedResponsibleId = partyId
            }
            if let startStr = r.startDate?.date, let date = dateFormatter.date(from: startStr) {
                startDate = date
            }
            if let endStr = r.completionDate?.date, let date = dateFormatter.date(from: endStr) {
                hasCompletionDate = true
                completionDate = date
            }
        } else if let defaultId = apiClient.defaultCheckoutUser?.id {
            selectedResponsibleId = defaultId
        }
    }

    private func save() async {
        isSaving = true
        defer { isSaving = false }

        let startStr = dateFormatter.string(from: startDate)
        let endStr = hasCompletionDate ? dateFormatter.string(from: completionDate) : nil
        let supplierIdOpt: Int? = selectedSupplierId == 0 ? nil : selectedSupplierId
        let costOpt: String? = NumberFormatHelpers.normalizeDecimalForAPI(cost)
        let urlOpt: String? = url.trimmingCharacters(in: .whitespaces).isEmpty ? nil : url.trimmingCharacters(in: .whitespaces)
        let notesOpt: String? = notes.trimmingCharacters(in: .whitespaces).isEmpty ? nil : notes
        let responsibleIdOpt: Int? = selectedResponsibleId == 0 ? nil : selectedResponsibleId

        // send both so the legacy string column stays populated too
        let typeIdOpt: Int? = usesTypeIds ? (selectedTypeId == 0 ? nil : selectedTypeId) : nil
        let typeStringOpt: String? = usesTypeIds
            ? apiClient.maintenanceTypes.first(where: { $0.id == selectedTypeId })?.name
            : selectedType

        let update = MaintenanceUpdateRequest(
            name: title,
            asset_maintenance_type: typeStringOpt,
            maintenance_type_id: typeIdOpt,
            supplier_id: supplierIdOpt,
            cost: costOpt,
            notes: notesOpt,
            url: urlOpt,
            responsible_party_id: responsibleIdOpt,
            start_date: startStr,
            completion_date: endStr,
            is_warranty: isWarranty,
            image_delete: (selectedImage == nil && removeExistingImage) ? 1 : nil
        )

        let savedId: Int?
        if let existing = record {
            savedId = await apiClient.updateMaintenance(
                id: existing.id,
                assetId: assetId,
                update: update,
                image: selectedImage,
                wasCompleted: existing.isCompleted
            )
        } else {
            let create = MaintenanceCreateRequest(
                asset_id: assetId,
                name: title,
                asset_maintenance_type: typeStringOpt,
                maintenance_type_id: typeIdOpt,
                supplier_id: supplierIdOpt,
                cost: costOpt,
                notes: notesOpt,
                url: urlOpt,
                responsible_party_id: responsibleIdOpt,
                start_date: startStr,
                completion_date: endStr,
                is_warranty: isWarranty
            )
            let created = await apiClient.createMaintenance(create, image: selectedImage)
            savedId = created ? 0 : nil
        }

        if savedId != nil {
            onSave(savedId ?? 0)
            dismiss()
        } else {
            errorMessage = apiClient.lastApiMessage ?? apiClient.errorMessage ?? L10n.string("error")
            showErrorAlert = true
        }
    }
}

File: Android/app/src/main/java/com/callandt/snipemobile/ui/maintenance/MaintenanceFormSheet.kt
package com.callandt.snipemobile.ui.maintenance

import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import com.callandt.snipemobile.data.api.UploadFile
import com.callandt.snipemobile.data.model.AssetMaintenance
import com.callandt.snipemobile.data.model.MaintenanceTypesMode
import com.callandt.snipemobile.ui.AppViewModel
import com.callandt.snipemobile.ui.asset.AssetFormSheetScaffold
import com.callandt.snipemobile.ui.asset.AssetFullScreenSheet
import com.callandt.snipemobile.ui.asset.AssetPhotoSection
import com.callandt.snipemobile.ui.asset.FormDateField
import com.callandt.snipemobile.ui.asset.FormSectionTitle
import com.callandt.snipemobile.ui.asset.PendingAssetImage
import com.callandt.snipemobile.ui.asset.formatApiDate
import com.callandt.snipemobile.ui.asset.normalizeDecimalForApi
import com.callandt.snipemobile.ui.components.PickerItem
import com.callandt.snipemobile.ui.components.SearchablePickerField
import com.callandt.snipemobile.ui.components.StringPickerField
import com.callandt.snipemobile.ui.util.L10n
import com.callandt.snipemobile.ui.util.resolveSnipeImageUrl
import com.callandt.snipemobile.ui.util.usersForNamePicker
import kotlinx.coroutines.launch
import java.util.Date

internal val legacyMaintenanceTypes = listOf(
    "Maintenance",
    "Repair",
    "PAT Test/Electrical",
    "Upgrade",
    "Hardware Support",
    "Software Support",
)

@Composable
fun AddMaintenanceSheet(
    viewModel: AppViewModel,
    initialAssetId: Int? = null,
    lockAssetSelection: Boolean = false,
    onDismiss: () -> Unit,
    onSaved: (Int) -> Unit = {},
) {
    MaintenanceFormSheet(
        viewModel = viewModel,
        title = L10n.string("add_maintenance"),
        saveLabel = L10n.string("save"),
        existing = null,
        initialAssetId = initialAssetId,
        lockAssetSelection = lockAssetSelection,
        onDismiss = onDismiss,
        onSaved = onSaved,
    )
}

@Composable
fun EditMaintenanceSheet(
    record: AssetMaintenance,
    viewModel: AppViewModel,
    onDismiss: () -> Unit,
    onSaved: (Int) -> Unit = {},
) {
    MaintenanceFormSheet(
        viewModel = viewModel,
        title = L10n.string("edit_maintenance"),
        saveLabel = L10n.string("save"),
        existing = record,
        initialAssetId = record.assetId,
        lockAssetSelection = true,
        onDismiss = onDismiss,
        onSaved = onSaved,
    )
}

@Composable
private fun MaintenanceFormSheet(
    viewModel: AppViewModel,
    title: String,
    saveLabel: String,
    existing: AssetMaintenance?,
    initialAssetId: Int?,
    lockAssetSelection: Boolean,
    onDismiss: () -> Unit,
    onSaved: (Int) -> Unit,
) {
    val assets by viewModel.assets.collectAsState()
    val users by viewModel.users.collectAsState()
    val currentUser by viewModel.currentUser.collectAsState()
    val suppliers by viewModel.suppliers.collectAsState()
    val maintenanceTypes by viewModel.apiClient.maintenanceTypes.collectAsState()
    val typesMode by viewModel.apiClient.maintenanceTypesMode.collectAsState()
    val lastApiMessage by viewModel.lastApiMessage.collectAsState()
    val scope = rememberCoroutineScope()
    val pickerUsers = remember(users, currentUser) { usersForNamePicker(users, currentUser) }

    var selectedAssetId by remember(existing?.id) { mutableIntStateOf(initialAssetId ?: 0) }
    var maintenanceTitle by remember(existing?.id) { mutableStateOf(existing?.decodedTitle.orEmpty()) }
    var selectedLegacyType by remember(existing?.id) {
        mutableStateOf(existing?.displayType ?: legacyMaintenanceTypes.first())
    }
    var selectedTypeId by remember(existing?.id) { mutableIntStateOf(0) }
    var selectedSupplierId by remember(existing?.id) { mutableIntStateOf(existing?.supplier?.id ?: 0) }
    var selectedUserId by remember(existing?.id) { mutableIntStateOf(existing?.responsibleParty?.id ?: 0) }
    var responsibleWasCleared by remember(existing?.id) { mutableStateOf(false) }
    var cost by remember(existing?.id) { mutableStateOf(existing?.cost.orEmpty()) }
    var url by remember(existing?.id) { mutableStateOf(existing?.url.orEmpty()) }
    var notes by remember(existing?.id) { mutableStateOf(existing?.decodedNotes.orEmpty()) }
    var isWarranty by remember(existing?.id) { mutableStateOf(existing?.isWarranty ?: false) }
    var startDateText by remember(existing?.id) {
        mutableStateOf(existing?.startDate?.date ?: formatApiDate(Date()))
    }
    var hasCompletionDate by remember(existing?.id) {
        mutableStateOf(!existing?.completionDate?.date.isNullOrEmpty())
    }
    var completionDateText by remember(existing?.id) {
        mutableStateOf(existing?.completionDate?.date ?: formatApiDate(Date()))
    }
    var pendingImage by remember(existing?.id) { mutableStateOf<PendingAssetImage?>(null) }
    var removeExistingImage by remember(existing?.id) { mutableStateOf(false) }
    var isSaving by remember { mutableStateOf(false) }
    var errorMessage by remember { mutableStateOf<String?>(null) }

    val isEditing = existing != null
    val usesTypeIds = typesMode == MaintenanceTypesMode.TypeIds
    val legacyOptions = remember(selectedLegacyType, existing?.displayType) {
        buildLegacyTypeOptions(selectedLegacyType, existing?.displayType)
    }
    val existingImageUrl = remember(existing?.image, existing?.updatedAt, viewModel.apiClient.baseUrl) {
        resolveSnipeImageUrl(
            viewModel.apiClient.baseUrl,
            existing?.image,
            existing?.updatedAt?.datetime ?: existing?.updatedAt?.date,
        )
    }

    LaunchedEffect(Unit) {
        if (suppliers.isEmpty()) viewModel.apiClient.fetchSuppliers()
        if (users.isEmpty()) viewModel.apiClient.fetchUsers()
        viewModel.apiClient.fetchMaintenanceTypes()
    }

    LaunchedEffect(maintenanceTypes, typesMode, existing?.id) {
        if (usesTypeIds && maintenanceTypes.isNotEmpty()) {
            applyTypeIdSelection(selectedTypeId = { selectedTypeId = it }, types = maintenanceTypes, record = existing)
        } else if (!usesTypeIds) {
            normalizeLegacyTypeSelection(selectedLegacyType) { selectedLegacyType = it }
        }
    }

    LaunchedEffect(users, existing?.id, responsibleWasCleared) {
        if (responsibleWasCleared) return@LaunchedEffect
        val preferredId = existing?.responsibleParty?.id ?: return@LaunchedEffect
        if (selectedUserId <= 0 && users.any { it.id == preferredId }) {
            selectedUserId = preferredId
        }
    }

    val typeIsValid = when (typesMode) {
        MaintenanceTypesMode.TypeIds -> maintenanceTypes.isNotEmpty() && selectedTypeId > 0
        MaintenanceTypesMode.Legacy -> selectedLegacyType.trim().isNotEmpty()
        MaintenanceTypesMode.Unknown -> false
    }
    val canSave = maintenanceTitle.trim().isNotEmpty() &&
        typeIsValid &&
        selectedAssetId > 0 &&
        !isSaving

    fun buildBody(): Map<String, Any?>? {
        val typeFields = resolveTypeFields(
            mode = typesMode,
            types = maintenanceTypes,
            selectedTypeId = selectedTypeId,
            selectedLegacyType = selectedLegacyType,
        ) ?: return null

        val body = mutableMapOf<String, Any?>(
            "name" to maintenanceTitle.trim(),
            "start_date" to startDateText.trim().take(10),
            "is_warranty" to isWarranty,
        )
        typeFields.first?.let { body["maintenance_type_id"] = it }
        typeFields.second?.let { body["asset_maintenance_type"] = it }
        normalizeDecimalForApi(cost)?.let { body["cost"] = it }
        notes.trim().takeIf { it.isNotEmpty() }?.let { body["notes"] = it }
        url.trim().takeIf { it.isNotEmpty() }?.let { body["url"] = it }
        if (selectedSupplierId > 0) {
            body["supplier_id"] = selectedSupplierId
        } else if (isEditing) {
            body["supplier_id"] = null
        }
        when {
            selectedUserId > 0 -> body["responsible_party_id"] = selectedUserId
            responsibleWasCleared && isEditing -> body["responsible_party_id"] = null
        }
        if (hasCompletionDate) {
            body["completion_date"] = completionDateText.trim().take(10)
        } else if (isEditing) {
            body["completion_date"] = null
        }
        if (!isEditing) body["asset_id"] = selectedAssetId
        if (removeExistingImage && pendingImage == null) body["image_delete"] = 1
        return body
    }

    AssetFullScreenSheet(onDismiss = { if (!isSaving) onDismiss() }) {
        AssetFormSheetScaffold(
            title = title,
            saveLabel = saveLabel,
            isSaving = isSaving,
            canSave = canSave,
            onDismiss = { if (!isSaving) onDismiss() },
            onSave = {
                isSaving = true
                scope.launch {
                    viewModel.apiClient.fetchMaintenanceTypes()
                    val body = buildBody()
                    if (body == null) {
                        errorMessage = lastApiMessage ?: L10n.string("error")
                        isSaving = false
                        return@launch
                    }
                    val imageUpload = pendingImage?.let { UploadFile("maintenance.jpg", it.mimeType, it.bytes) }
                    val savedId = if (isEditing) {
                        val record = existing!!
                        val assetId = record.assetId ?: selectedAssetId
                        viewModel.apiClient.updateMaintenance(
                            id = record.id,
                            assetId = assetId,
                            body = body,
                            image = imageUpload,
                            imageDelete = removeExistingImage && imageUpload == null,
                            wasCompleted = record.isCompleted,
                        ).also { id ->
                            if (id == null) errorMessage = lastApiMessage ?: L10n.string("mgmt_save_failed")
                        }
                    } else {
                        viewModel.apiClient.createMaintenanceReturningId(body, imageUpload).also { id ->
                            if (id == null) errorMessage = lastApiMessage ?: L10n.string("create_failed")
                        }
                    }
                    isSaving = false
                    if (savedId != null) {
                        viewModel.syncInBackground()
                        onSaved(savedId)
                        onDismiss()
                    }
                }
            },
        ) {
            if (!isEditing && !lockAssetSelection) {
                FormSectionTitle(L10n.string("asset"))
                SearchablePickerField(
                    label = L10n.fieldLabel("asset", required = true),
                    items = assets.map { asset ->
                        PickerItem(
                            id = asset.id,
                            title = asset.decodedName,
                            subtitle = asset.decodedAssetTag,
                        )
                    },
                    selectedId = selectedAssetId.takeIf { it > 0 },
                    onSelected = { selectedAssetId = it.id },
                )
            }

            FormSectionTitle(L10n.string("general"))
            OutlinedTextField(
                value = maintenanceTitle,
                onValueChange = { maintenanceTitle = it },
                label = { Text(L10n.fieldLabel("name", required = true)) },
                modifier = Modifier.fillMaxWidth(),
            )
            when (typesMode) {
                MaintenanceTypesMode.Unknown -> Text(L10n.string("loading"), modifier = Modifier.fillMaxWidth())
                MaintenanceTypesMode.TypeIds -> {
                    if (maintenanceTypes.isNotEmpty() && maintenanceTypes.any { it.id == selectedTypeId }) {
                        SearchablePickerField(
                            label = L10n.fieldLabel("maintenance_type", required = true),
                            items = maintenanceTypes.map { PickerItem(it.id, it.decodedName) },
                            selectedId = selectedTypeId.takeIf { it > 0 },
                            onSelected = { selectedTypeId = it.id },
                        )
                    } else {
                        Text(L10n.string("loading"), modifier = Modifier.fillMaxWidth())
                    }
                }
                MaintenanceTypesMode.Legacy -> {
                    StringPickerField(
                        label = L10n.fieldLabel("maintenance_type", required = true),
                        options = legacyOptions.map { it to it },
                        selectedValue = selectedLegacyType,
                        onSelected = { selectedLegacyType = it },
                    )
                }
            }

            FormSectionTitle(L10n.string("responsible_party"))
            if (users.isNotEmpty()) {
                SearchablePickerField(
                    label = L10n.string("responsible_party"),
                    items = pickerUsers.map { PickerItem(it.id, it.decodedName, it.decodedEmail) },
                    selectedId = selectedUserId.takeIf { it > 0 },
                    placeholder = L10n.string("none"),
                    allowClear = selectedUserId > 0,
                    onClear = {
                        selectedUserId = 0
                        responsibleWasCleared = true
                    },
                    onSelected = {
                        selectedUserId = it.id
                        responsibleWasCleared = false
                    },
                )
            } else {
                Text(L10n.string("loading"), modifier = Modifier.fillMaxWidth())
            }

            FormSectionTitle(L10n.string("dates"))
            FormDateField(
                label = L10n.string("start_date"),
                dateText = startDateText,
                onDateTextChange = { startDateText = it },
            )
            Row(
                modifier = Modifier.fillMaxWidth(),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text(L10n.string("set_completion_date"), modifier = Modifier.weight(1f))
                Switch(checked = hasCompletionDate, onCheckedChange = { hasCompletionDate = it })
            }
            if (hasCompletionDate) {
                FormDateField(
                    label = L10n.string("completion_date"),
                    dateText = completionDateText,
                    onDateTextChange = { completionDateText = it },
                )
            }

            FormSectionTitle(L10n.string("financial"))
            OutlinedTextField(
                value = cost,
                onValueChange = { cost = it },
                label = { Text(L10n.string("cost")) },
                modifier = Modifier.fillMaxWidth(),
            )
            if (suppliers.isNotEmpty()) {
                SearchablePickerField(
                    label = L10n.string("supplier_optional"),
                    items = suppliers.map { PickerItem(it.id, it.decodedName) },
                    selectedId = selectedSupplierId.takeIf { it > 0 },
                    placeholder = L10n.string("none"),
                    allowClear = selectedSupplierId > 0,
                    onClear = { selectedSupplierId = 0 },
                    onSelected = { selectedSupplierId = it.id },
                )
            }

            Row(
                modifier = Modifier.fillMaxWidth(),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text(L10n.string("is_warranty"), modifier = Modifier.weight(1f))
                Switch(checked = isWarranty, onCheckedChange = { isWarranty = it })
            }

            FormSectionTitle(L10n.string("url"))
            OutlinedTextField(
                value = url,
                onValueChange = { url = it },
                label = { Text(L10n.string("url")) },
                modifier = Modifier.fillMaxWidth(),
                singleLine = true,
            )

            AssetPhotoSection(
                pendingImage = pendingImage,
                onPendingImageChange = { pendingImage = it },
                existingImageUrl = existingImageUrl,
                removeExistingImage = removeExistingImage,
                onRemoveExistingImageChange = { removeExistingImage = it },
            )

            FormSectionTitle(L10n.string("notes"))
            OutlinedTextField(
                value = notes,
                onValueChange = { notes = it },
                label = { Text(L10n.string("notes")) },
                modifier = Modifier.fillMaxWidth(),
                minLines = 4,
            )
        }
    }

    errorMessage?.let { message ->
        AlertDialog(
            onDismissRequest = { errorMessage = null },
            title = { Text(L10n.string("error")) },
            text = { Text(message) },
            confirmButton = {
                TextButton(onClick = { errorMessage = null }) { Text(L10n.string("ok")) }
            },
        )
    }
}

internal fun buildLegacyTypeOptions(selectedType: String, recordType: String?): List<String> {
    val options = legacyMaintenanceTypes.toMutableList()
    listOfNotNull(selectedType, recordType)
        .filter { it.isNotEmpty() && !options.contains(it) }
        .forEach { options.add(it) }
    return options
}

internal fun normalizeLegacyTypeSelection(current: String, onUpdate: (String) -> Unit) {
    val options = buildLegacyTypeOptions(current, null)
    if (!options.contains(current) && options.isNotEmpty()) onUpdate(options.first())
}

internal fun applyTypeIdSelection(
    selectedTypeId: (Int) -> Unit,
    types: List<com.callandt.snipemobile.data.model.MaintenanceType>,
    record: AssetMaintenance?,
) {
    if (types.isEmpty()) return
    record?.let { r ->
        val target = (r.maintenanceType ?: r.assetMaintenanceType)?.lowercase()
        if (!target.isNullOrEmpty()) {
            types.firstOrNull {
                it.name.lowercase() == target || it.decodedName.lowercase() == target
            }?.let {
                selectedTypeId(it.id)
                return
            }
        }
    }
    types.firstOrNull()?.let { selectedTypeId(it.id) }
}

internal fun resolveTypeFields(
    mode: MaintenanceTypesMode,
    types: List<com.callandt.snipemobile.data.model.MaintenanceType>,
    selectedTypeId: Int,
    selectedLegacyType: String,
): Pair<Int?, String?>? {
    return when (mode) {
        MaintenanceTypesMode.TypeIds -> {
            if (selectedTypeId <= 0) null
            else {
                val name = types.firstOrNull { it.id == selectedTypeId }?.name
                selectedTypeId to name
            }
        }
        MaintenanceTypesMode.Legacy -> {
            val trimmed = selectedLegacyType.trim()
            if (trimmed.isEmpty()) null else null to trimmed
        }
        MaintenanceTypesMode.Unknown -> null
    }
}
